Mathlib/Order/Hom/Lex.lean

```diff
@@ -0,0 +1,141 @@
+/-
+Copyright (c) 2025 Violeta Hernández Palacios. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Violeta Hernández Palacios
+-/
+import Mathlib.Data.Sum.Order
+import Mathlib.Order.Hom.Set
+import Mathlib.Order.RelIso.Set
+
+/-!
+# Lexicographic order and order isomorphisms
+
+The main result in this file is the following: if `α` is a linear order and `x : α`, then `α` is
+order isomorphic to either of `Iio x ⊕ₗ Ici x` or `Iic x ⊕ₗ Ioi x`.
+-/
+
+open Set
+
+variable {α : Type*}
+
+/-! ### Relation isomorphism -/
+
+namespace RelIso
+
+variable {r : α → α → Prop} {x y : α} [IsTrans α r] [IsTrichotomous α r] [DecidableRel r]
+
+variable (r x) in
+/-- A relation is isomorphic to the lexicographic sum of elements less than `x` and elements not
+less than `x`. -/
+-- The explicit type signature stops `simp` from complaining.
+def sumLexComplLeft :
+    @Sum.Lex {y // r y x} {y // ¬ r y x} (Subrel r {y | r y x}) (Subrel r {y | ¬ r y x}) ≃r r where
+  toEquiv := Equiv.sumCompl (r · x)
+  map_rel_iff' := by
+    rintro (⟨a, ha⟩ | ⟨a, ha⟩) (⟨b, hb⟩ | ⟨b, hb⟩)
+    · simp
+    · simpa using trans_trichotomous_right ha hb
+    · simpa using fun h ↦ ha <| trans h hb
+    · simp
+
+@[simp]
+theorem sumLexComplLeft_apply (a) : sumLexComplLeft r x a = Equiv.sumCompl (r · x) a :=
+  rfl
+
+@[simp]
+theorem sumLexComplLeft_symm_apply (a) : sumLexComplLeft r x a = Equiv.sumCompl (r · x) a :=
+  rfl
+
+variable (r x) in
+/-- A relation is isomorphic to the lexicographic sum of elements not greater than `x` and elements
+greater than `x`. -/
+-- The explicit type signature stops `simp` from complaining.
+def sumLexComplRight :
+    @Sum.Lex {y // ¬ r x y} {y // r x y} (Subrel r {y | ¬ r x y}) (Subrel r {y | r x y}) ≃r r where
+  toEquiv := (Equiv.sumComm _ _).trans <| Equiv.sumCompl (r x ·)
+  map_rel_iff' := by
+    rintro (⟨a, ha⟩ | ⟨a, ha⟩) (⟨b, hb⟩ | ⟨b, hb⟩)
+    · simp
+    · simpa using trans_trichotomous_left ha hb
+    · simpa using fun h ↦ hb <| trans ha h
+    · simp
+
+@[simp]
+theorem sumLexComplRight_apply (a) : sumLexComplRight r x a = Equiv.sumCompl (r x ·) a.swap :=
+  rfl
+
+@[simp]
+theorem sumLexComplRight_symm_apply (a) : sumLexComplRight r x a = Equiv.sumCompl (r x ·) a.swap :=
+  rfl
+
+end RelIso
+
+/-! ### Order isomorphism -/
+
+namespace OrderIso
+
+variable [LinearOrder α] {x y : α}
+
+variable (x) in
+/-- A linear order is isomorphic to the lexicographic sum of elements less than `x` and elements
+greater or equal to `x`. -/
+def sumLexIioIci : Iio x ⊕ₗ Ici x ≃o α :=
+  (sumLexCongr (refl _) (setCongr (Ici x) {y | ¬ y < x} (by ext; simp))).trans <|
+    ofRelIsoLT (RelIso.sumLexComplLeft (· < ·) x)
+
+@[simp]
+theorem sumLexIioIci_apply_inl (a : Iio x) : sumLexIioIci x (Sum.inl a) = a :=
+  rfl
+
+@[simp]
+theorem sumLexIioIci_apply_inr (a : Ici x) : sumLexIioIci x (Sum.inr a) = a :=
+  rfl
```

Comment:
```suggestion
theorem sumLexIioIci_apply_inl (a : Iio x) : sumLexIioIci x (ofLex <| Sum.inl a) = a :=
  rfl

@[simp]
theorem sumLexIioIci_apply_inr (a : Ici x) : sumLexIioIci x (ofLex <| Sum.inr a) = a :=
  rfl
```

Reply:
That suggestion doesn't make any sense to me, the type of this bundled map is  `Lex (Sum (Iio x) (Ici x)) ≃o α`, so when applying the map to a `Sum` you have to call `toLex` first.

(I realize I made a type above too...)

Reply:
Oh nevermind, I see why what I was saying is nonsense.